Lib/fontTools/subset/svg.py

```diff
@@ -0,0 +1,235 @@
+        ".//svg:*[starts-with(@xlink:href, '#glyph')]", namespaces=NAMESPACES
+    )
+
+
+def group_elements_by_id(tree: etree.Element) -> Dict[str, etree.Element]:
+    return {el.attrib["id"]: el for el in find_svg_elements_with_id(tree)}
+
+
+def parse_css_declarations(style_attr: str) -> Dict[str, str]:
+    # https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/style
+    # https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#css_declarations
+    result = {}
+    for declaration in style_attr.split(";"):
+        if declaration.count(":") == 1:
+            property_name, value = declaration.split(":")
+            property_name = property_name.strip()
+            result[property_name] = value.strip()
+        elif declaration.strip():
+            raise ValueError(f"Invalid CSS declaration syntax: {declaration}")
+    return result
+
+
+def iter_referenced_ids(tree: etree.Element) -> Iterator[str]:
+    # Yield all the ids that can be reached via references from within this element tree
+    for el in chain([tree], find_svg_elements_with_references(tree)):
+        if XLINK_HREF in el.attrib:
+            href = el.attrib[XLINK_HREF]
+            if href.startswith("#"):
+                ref_id = href[1:]
+                assert ref_id
+                yield ref_id
+
+        attrs = el.attrib
+        if "style" in attrs:
+            attrs = {**attrs, **parse_css_declarations(el.attrib["style"])}
+        for attr in ("fill", "clip-path"):
+            if attr in attrs:
+                value = attrs[attr]
+                if value.startswith("url(#") and value.endswith(")"):
+                    ref_id = value[5:-1]
+                    assert ref_id
+                    yield ref_id
+
+
+def closure_element_ids(
+    elements: Dict[str, etree.Element], element_ids: Set[str]
+) -> None:
+    # Expand the initial subset of element ids to include ids that can be reached
+    # via references from the initial set.
+    unvisited = element_ids
+    while unvisited:
+        referenced: Set[str] = set()
+        for el_id in unvisited:
+            if el_id not in elements:
+                # ignore dangling reference; not our job to validate svg
+                continue
+            referenced.update(iter_referenced_ids(elements[el_id]))
+        referenced -= element_ids
+        element_ids.update(referenced)
+        unvisited = referenced
+
+
+def subset_elements(el: etree.Element, ids: Set[str]) -> bool:
+    # Keep elements if their id is in the subset, or any of their children's id is.
+    # Drop elements whose id is not in the subset, and either have no children,
+    # or all their children are being dropped.
+    el_id = el.attrib.get("id")
+    if el_id is not None and el_id in ids:
+        # if id is in the set, don't recurse; keep whole subtree
+        return True
+    keep = False
+    for e in el:
+        keep |= subset_elements(e, ids)
+    if keep:
+        return True
```

Comment:
Do you specifically want to run subset_elements for every child? - if so, plz comment why. If not perhaps possible to make terser

```
if not any(subset_elements(e, retained_ids) for e in el):
  # remove parent
return el.getparent() is not None  # it's only none if we dropped
```

Reply:
yes, I do want to run for every child. I don't want to do any(...) as that would stop at the first match. The subset_elements has the side effect of removing elements that are left empty

Reply:
You could use al then I think?

Reply:
`all` would also short-circuit (as soon as one is False). Actually I can use a list comprehension to run subset_elements on all children and avoid any() to short-circuit. I'll do that